File: ChatClientVB/ViewModels/MainWindowViewModel.vb
Imports System.Collections.ObjectModel
Imports System.IO
Imports System.Drawing
Imports System.Reactive.Linq

Public Class MainWindowViewModel
    Inherits ViewModelBase

    Private chatService As IChatService
    Private dialogService As IDialogService
    Private ctxTaskFactory As TaskFactory
    Private Const MAX_IMAGE_WIDTH As Integer = 150
    Private Const MAX_IMAGE_HEIGHT As Integer = 150

    Private _userName As String
    Public Property UserName As String
        Get
            Return _userName
        End Get
        Set(value As String)
            _userName = value
            OnPropertyChanged()
        End Set
    End Property

    Private _profilePic As String
    Public Property ProfilePic As String
        Get
            Return _profilePic
        End Get
        Set(value As String)
            _profilePic = value
            OnPropertyChanged()
        End Set
    End Property

    Private _participants As New ObservableCollection(Of Participant)
    Public Property Participants As ObservableCollection(Of Participant)
        Get
            Return _participants
        End Get
        Set(value As ObservableCollection(Of Participant))
            _participants = value
            OnPropertyChanged()
        End Set
    End Property

    Private _selectedParticipant As Participant
    Public Property SelectedParticipant As Participant
        Get
            Return _selectedParticipant
        End Get
        Set(value As Participant)
            _selectedParticipant = value
            If SelectedParticipant.HasSentNewMessage Then SelectedParticipant.HasSentNewMessage = False
            OnPropertyChanged()
        End Set
    End Property

    Private _userMode As UserModes
    Public Property UserMode As UserModes
        Get
            Return _userMode
        End Get
        Set(value As UserModes)
            _userMode = value
            OnPropertyChanged()
        End Set
    End Property

    Private _textMessage As String
    Public Property TextMessage As String
        Get
            Return _textMessage
        End Get
        Set(value As String)
            _textMessage = value
            OnPropertyChanged()
        End Set
    End Property

    Private _isConnected As Boolean
    Public Property IsConnected As Boolean
        Get
            Return _isConnected
        End Get
        Set(value As Boolean)
            _isConnected = value
            OnPropertyChanged()
        End Set
    End Property

    Private _isLoggedIn As Boolean
    Public Property IsLoggedIn As Boolean
        Get
            Return _isLoggedIn
        End Get
        Set(value As Boolean)
            _isLoggedIn = value
            OnPropertyChanged()
        End Set
    End Property

#Region "Connect Command"
    Private _connectCommand As ICommand
    Public ReadOnly Property ConnectCommand As ICommand
        Get
            Return If(_connectCommand, New RelayCommandAsync(Function() Connect()))
        End Get
    End Property

    Private Async Function Connect() As Task(Of Boolean)
        Try
            Await chatService.ConnectAsync()
            IsConnected = True
            Return True
        Catch ex As Exception
            Return False
        End Try
    End Function
#End Region

#Region "Login Command"
    Private _loginCommand As ICommand
    Public ReadOnly Property LoginCommand As ICommand
        Get
            Return If(_loginCommand, New RelayCommandAsync(Function() Login(), AddressOf CanLogin))
        End Get
    End Property

    Private Async Function Login() As Task(Of Boolean)
        Try
            Dim users As List(Of User)
            users = Await chatService.LoginAsync(_userName, Avatar())
            If users IsNot Nothing Then
                users.ForEach(Sub(u) Participants.Add(New Participant With {.Name = u.Name, .Photo = u.Photo}))
                UserMode = UserModes.Chat
                IsLoggedIn = True
                Return True
            Else
                dialogService.ShowNotification("Username is already in use")
                Return False
            End If
        Catch ex As Exception
            Return False
        End Try
    End Function

    Private Function CanLogin() As Boolean
        Return Not String.IsNullOrEmpty(UserName) AndAlso UserName.Length >= 2 AndAlso IsConnected
    End Function
#End Region

#Region "Logout Command"
    Private _logoutCommand As ICommand
    Public ReadOnly Property LogoutCommand As ICommand
        Get
            Return If(_logoutCommand, New RelayCommandAsync(Function() Logout(), AddressOf CanLogout))
        End Get
    End Property

    Private Async Function Logout() As Task(Of Boolean)
        Try
            Await chatService.LogoutAsync()
            UserMode = UserModes.Login
            Return True
        Catch ex As Exception
            Return False
        End Try
    End Function

    Private Function CanLogout() As Boolean
        Return IsConnected AndAlso IsLoggedIn
    End Function
#End Region

#Region "Typing Command"
    Private _typingCommand As ICommand
    Public ReadOnly Property TypingCommand As ICommand
        Get
            Return If(_typingCommand, New RelayCommandAsync(Function() Typing(), AddressOf CanUseTypingCommand))
        End Get
    End Property

    Private Async Function Typing() As Task(Of Boolean)
        Try
            Await chatService.TypingAsync(SelectedParticipant.Name)
            Return True
        Catch ex As Exception
            Return False
        End Try
    End Function

    Private Function CanUseTypingCommand() As Boolean
        Return SelectedParticipant IsNot Nothing AndAlso SelectedParticipant.IsLoggedIn
    End Function
#End Region

#Region "Send Text Message Command"
    Private _sendTextMessageCommand As ICommand
    Public ReadOnly Property SendTextMessageCommand As ICommand
        Get
            Return If(_sendTextMessageCommand, New RelayCommandAsync(Function() SendTextMessage(),
                                                                     AddressOf CanSendTextMessage))
        End Get
    End Property

    Private Async Function SendTextMessage() As Task(Of Boolean)
        Try
            Dim recepient = _selectedParticipant.Name
            Await chatService.SendUnicastMessageAsync(recepient, _textMessage)
            Return True
        Catch ex As Exception
            Return False
        Finally
            Dim msg As New ChatMessage With {.Author = UserName, .Message = _textMessage,
                .Time = DateTime.Now, .IsOriginNative = True}
            SelectedParticipant.Chatter.Add(msg)
            TextMessage = String.Empty
        End Try
    End Function

    Private Function CanSendTextMessage() As Boolean
        Return Not String.IsNullOrEmpty(TextMessage) AndAlso IsConnected AndAlso
            _selectedParticipant IsNot Nothing AndAlso _selectedParticipant.IsLoggedIn
    End Function
#End Region

#Region "Send Picture Message Command"
    Private _sendImageMessageCommand As ICommand
    Public ReadOnly Property SendImageMessageCommand As ICommand
        Get
            Return If(_sendImageMessageCommand, New RelayCommandAsync(Function() SendImageMessage(),
                                                                     AddressOf CanSendImageMessage))
        End Get
    End Property

    Private Async Function SendImageMessage() As Task(Of Boolean)
        Dim pic = dialogService.OpenFile("Select image file", "Images (*.jpg;*.png)|*.jpg;*.png")
        If String.IsNullOrEmpty(pic) Then Return False

        Dim img = Await Task.Run(Function() File.ReadAllBytes(pic))

        Try
            Dim recepient = _selectedParticipant.Name
            Await chatService.SendUnicastMessageAsync(recepient, img)
            Return True
        Catch ex As Exception
            Return False
        Finally
            Dim msg As New ChatMessage With {.Author = UserName, .Picture = pic, .Time = DateTime.Now, .IsOriginNative = True}
            SelectedParticipant.Chatter.Add(msg)
        End Try
    End Function

    Private Function CanSendImageMessage() As Boolean
        Return IsConnected AndAlso _selectedParticipant IsNot Nothing AndAlso _selectedParticipant.IsLoggedIn
    End Function
#End Region

#Region "Select Profile Picture Command"
    Private _selectProfilePicCommand As ICommand
    Public ReadOnly Property SelectPhotoCommand As ICommand
        Get
            Return If(_selectProfilePicCommand, New RelayCommand(AddressOf SelectProfilePic))
        End Get
    End Property

    Private Sub SelectProfilePic()
        Dim pic = dialogService.OpenFile("Select image file", "Images (*.jpg;*.png)|*.jpg;*.png")
        If Not String.IsNullOrEmpty(pic) Then
            Dim img = Image.FromFile(pic)
            If img.Width > MAX_IMAGE_WIDTH OrElse img.Height > MAX_IMAGE_HEIGHT Then
                dialogService.ShowNotification($"Image size should be {MAX_IMAGE_WIDTH} x {MAX_IMAGE_HEIGHT} or less.")
                Exit Sub
            End If
            ProfilePic = pic
        End If
    End Sub
#End Region

#Region "Open Image Command"
    Private _openImageCommand As ICommand
    Public ReadOnly Property OpenImageCommand As ICommand
        Get
            Return If(_openImageCommand, New RelayCommand(Of ChatMessage)(Sub(m) OpenImage(m)))
        End Get
    End Property

    Private Sub OpenImage(ByVal msg As ChatMessage)
        Dim img = msg.Picture
        If (String.IsNullOrEmpty(img) OrElse Not File.Exists(img)) Then Exit Sub
        Process.Start(img)
    End Sub
#End Region

#Region "Event handlers"
    Private Sub NewTextMessage(name As String, msg As String, mt As MessageType)
        If mt = MessageType.Unicast Then
            Dim cm As New ChatMessage With {.Author = name, .Message = msg, .Time = DateTime.Now}
            Dim sender = _participants.Where(Function(u) String.Equals(u.Name, name)).FirstOrDefault
            ctxTaskFactory.StartNew(Sub() sender.Chatter.Add(cm)).Wait()

            If Not (SelectedParticipant IsNot Nothing AndAlso sender.Name.Equals(SelectedParticipant.Name)) Then
                ctxTaskFactory.StartNew(Sub() sender.HasSentNewMessage = True).Wait()
            End If
        End If
    End Sub

    Private Sub NewImageMessage(name As String, pic As Byte(), mt As MessageType)
        If mt = MessageType.Unicast Then
            Dim imgsDirectory = Path.Combine(Environment.CurrentDirectory, "Image Messages")
            If Not Directory.Exists(imgsDirectory) Then Directory.CreateDirectory(imgsDirectory)

            Dim imgsCount = Directory.EnumerateFiles(imgsDirectory).Count() + 1
            Dim imgPath = Path.Combine(imgsDirectory, $"IMG_{imgsCount}.jpg")

            Dim converter As New ImageConverter
            Using img As Image = CType(converter.ConvertFrom(pic), Image)
                img.Save(imgPath)
            End Using

            Dim cm As New ChatMessage With {.Author = name, .Picture = imgPath, .Time = DateTime.Now}
            Dim sender = _participants.Where(Function(u) String.Equals(u.Name, name)).FirstOrDefault
            ctxTaskFactory.StartNew(Sub() sender.Chatter.Add(cm)).Wait()

            If Not (SelectedParticipant IsNot Nothing AndAlso sender.Name.Equals(SelectedParticipant.Name)) Then
                ctxTaskFactory.StartNew(Sub() sender.HasSentNewMessage = True).Wait()
            End If
        End If
    End Sub

    Private Sub ParticipantLogin(ByVal u As User)
        Dim ptp = Participants.FirstOrDefault(Function(p) String.Equals(p.Name, u.Name))
        If _isLoggedIn AndAlso ptp Is Nothing Then
            ctxTaskFactory.StartNew(Sub() Participants.Add(New Participant With {.Name = u.Name, .Photo = u.Photo})).Wait()
        End If
    End Sub

    Private Sub ParticipantDisconnection(ByVal name As String)
        Dim person = Participants.Where(Function(p) String.Equals(p.Name, name)).FirstOrDefault
        If person IsNot Nothing Then person.IsLoggedIn = False
    End Sub

    Private Sub ParticipantReconnection(ByVal name As String)
        Dim person = Participants.Where(Function(p) String.Equals(p.Name, name)).FirstOrDefault
        If person IsNot Nothing Then person.IsLoggedIn = True
    End Sub

    Private Sub Reconnecting()
        IsConnected = False
        IsLoggedIn = False
    End Sub

    Private Async Sub Reconnected()
        Dim pic = Avatar()
        If Not String.IsNullOrEmpty(_userName) Then Await chatService.LoginAsync(_userName, pic)
        IsConnected = True
        IsLoggedIn = True
    End Sub

    Private Async Sub Disconnected()
        Dim connectionTask = chatService.ConnectAsync()
        Await connectionTask.ContinueWith(Sub(t)
                                              If Not t.IsFaulted Then
                                                  IsConnected = True
                                                  chatService.LoginAsync(_userName, Avatar()).Wait()
                                                  IsLoggedIn = True
                                              End If
                                          End Sub)
    End Sub

    Private Sub ParticipantTyping(ByVal name As String)
        Dim person = Participants.Where(Function(p) String.Equals(p.Name, name)).FirstOrDefault()
        If person IsNot Nothing AndAlso Not person.IsTyping Then
            person.IsTyping = True
            Observable.Timer(TimeSpan.FromMilliseconds(1500)).Subscribe(Sub(t) person.IsTyping = False)
        End If
    End Sub
#End Region

    Private Function Avatar() As Byte()
        Dim pic As Byte() = Nothing
        If Not String.IsNullOrEmpty(_profilePic) Then pic = File.ReadAllBytes(_profilePic)
        Return pic
    End Function

    Public Sub New(chatSvc As IChatService, diagSvc As IDialogService)
        dialogService = diagSvc
        chatService = chatSvc
        AddHandler chatSvc.NewTextMessage, AddressOf NewTextMessage
        AddHandler chatSvc.NewImageMessage, AddressOf NewImageMessage
        AddHandler chatSvc.ParticipantLoggedIn, AddressOf ParticipantLogin
        AddHandler chatSvc.ParticipantLoggedOut, AddressOf ParticipantDisconnection
        AddHandler chatSvc.ParticipantDisconnected, AddressOf ParticipantDisconnection
        AddHandler chatSvc.ParticipantReconnected, AddressOf ParticipantReconnection
        AddHandler chatSvc.ParticipantTyping, AddressOf ParticipantTyping
        AddHandler chatSvc.ConnectionReconnecting, AddressOf Reconnecting
        AddHandler chatSvc.ConnectionReconnected, AddressOf Reconnected
        AddHandler chatSvc.ConnectionClosed, AddressOf Disconnected

        ctxTaskFactory = New TaskFactory(TaskScheduler.FromCurrentSynchronizationContext)
    End Sub
End Class

File: ChatClientCS/ViewModels/MainWindowViewModel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.Collections.ObjectModel;
using System.IO;
using System.Drawing;
using ChatClientCS.Services;
using ChatClientCS.Enums;
using ChatClientCS.Models;
using ChatClientCS.Commands;
using System.Windows.Input;
using System.Windows.Media;
using System.Diagnostics;
using System.Reactive.Linq;

namespace ChatClientCS.ViewModels
{
    public class MainWindowViewModel : ViewModelBase
    {
        private IChatService chatService;
        private IDialogService dialogService;
        private TaskFactory ctxTaskFactory;
        private const int MAX_IMAGE_WIDTH = 150;
        private const int MAX_IMAGE_HEIGHT = 150;

        private string _userName;
        public string UserName
        {
            get { return _userName; }
            set
            {
                _userName = value;
                OnPropertyChanged();
            }
        }

        private string _profilePic;
        public string ProfilePic
        {
            get { return _profilePic; }
            set
            {
                _profilePic = value;
                OnPropertyChanged();
            }
        }

        private ObservableCollection<Participant> _participants = new ObservableCollection<Participant>();
        public ObservableCollection<Participant> Participants
        {
            get { return _participants; }
            set
            {
                _participants = value;
                OnPropertyChanged();
            }
        }

        private Participant _selectedParticipant;
        public Participant SelectedParticipant
        {
            get { return _selectedParticipant; }
            set
            {
                _selectedParticipant = value;
                if (SelectedParticipant.HasSentNewMessage) SelectedParticipant.HasSentNewMessage = false;
                OnPropertyChanged();
            }
        }

        private UserModes _userMode;
        public UserModes UserMode
        {
            get { return _userMode; }
            set
            {
                _userMode = value;
                OnPropertyChanged();
            }
        }

        private string _textMessage;
        public string TextMessage
        {
            get { return _textMessage; }
            set
            {
                _textMessage = value;
                OnPropertyChanged();
            }
        }

        private bool _isConnected;
        public bool IsConnected
        {
            get { return _isConnected; }
            set
            {
                _isConnected = value;
                OnPropertyChanged();
            }
        }

        private bool _isLoggedIn;
        public bool IsLoggedIn
        {
            get { return _isLoggedIn; }
            set
            {
                _isLoggedIn = value;
                OnPropertyChanged();
            }
        }

        #region Connect Command
        private ICommand _connectCommand;
        public ICommand ConnectCommand
        {
            get
            {
                return _connectCommand ?? (_connectCommand = new RelayCommandAsync(() => Connect()));
            }
        }

        private async Task<bool> Connect()
        {
            try
            {
                await chatService.ConnectAsync();
                IsConnected = true;
                return true;
            }
            catch (Exception) { return false; }
        }
        #endregion

        #region Login Command
        private ICommand _loginCommand;
        public ICommand LoginCommand
        {
            get
            {
                return _loginCommand ?? (_loginCommand =
                    new RelayCommandAsync(() => Login(), (o) => CanLogin()));
            }
        }

        private async Task<bool> Login()
        {
            try
            {
                List<User> users = new List<User>();
                users = await chatService.LoginAsync(_userName, Avatar());
                if (users != null)
                {
                    users.ForEach(u => Participants.Add(new Participant { Name = u.Name, Photo = u.Photo }));
                    UserMode = UserModes.Chat;
                    IsLoggedIn = true;
                    return true;
                }
                else
                {
                    dialogService.ShowNotification("Username is already in use");
                    return false;
                }

            }
            catch (Exception) { return false; }
        }

        private bool CanLogin()
        {
            return !string.IsNullOrEmpty(UserName) && UserName.Length >= 2 && IsConnected;
        }
        #endregion

        #region Logout Command
        private ICommand _logoutCommand;
        public ICommand LogoutCommand
        {
            get
            {
                return _logoutCommand ?? (_logoutCommand =
                    new RelayCommandAsync(() => Logout(), (o) => CanLogout()));
            }
        }

        private async Task<bool> Logout()
        {
            try
            {
                await chatService.LogoutAsync();
                UserMode = UserModes.Login;
                return true;
            }
            catch (Exception) { return false; }
        }

        private bool CanLogout()
        {
            return IsConnected && IsLoggedIn;
        }
        #endregion

        #region Typing Command
        private ICommand _typingCommand;
        public ICommand TypingCommand
        {
            get
            {
                return _typingCommand ?? (_typingCommand =
                    new RelayCommandAsync(() => Typing(), (o) => CanUseTypingCommand()));
            }
        }

        private async Task<bool> Typing()
        {
            try
            {
                await chatService.TypingAsync(SelectedParticipant.Name);
                return true;
            }
            catch (Exception) { return false; }
        }

        private bool CanUseTypingCommand()
        {
            return (SelectedParticipant != null && SelectedParticipant.IsLoggedIn);
        }
        #endregion

        #region Send Text Message Command
        private ICommand _sendTextMessageCommand;
        public ICommand SendTextMessageCommand
        {
            get
            {
                return _sendTextMessageCommand ?? (_sendTextMessageCommand =
                    new RelayCommandAsync(() => SendTextMessage(), (o) => CanSendTextMessage()));
            }
        }

        private async Task<bool> SendTextMessage()
        {
            try
            {
                var recepient = _selectedParticipant.Name;
                await chatService.SendUnicastMessageAsync(recepient, _textMessage);
                return true;
            }
            catch (Exception) { return false; }
            finally
            {
                ChatMessage msg = new ChatMessage
                {
                    Author = UserName,
                    Message = _textMessage,
                    Time = DateTime.Now,
                    IsOriginNative = true
                };
                SelectedParticipant.Chatter.Add(msg);
                TextMessage = string.Empty;
            }
        }

        private bool CanSendTextMessage()
        {
            return (!string.IsNullOrEmpty(TextMessage) && IsConnected &&
                _selectedParticipant != null && _selectedParticipant.IsLoggedIn);
        }
        #endregion

        #region Send Picture Message Command
        private ICommand _sendImageMessageCommand;
        public ICommand SendImageMessageCommand
        {
            get
            {
                return _sendImageMessageCommand ?? (_sendImageMessageCommand =
                    new RelayCommandAsync(() => SendImageMessage(), (o) => CanSendImageMessage()));
            }
        }

        private async Task<bool> SendImageMessage()
        {
            var pic = dialogService.OpenFile("Select image file", "Images (*.jpg;*.png)|*.jpg;*.png");
            if (string.IsNullOrEmpty(pic)) return false;

            var img = await Task.Run(() => File.ReadAllBytes(pic));

            try
            {
                var recepient = _selectedParticipant.Name;
                await chatService.SendUnicastMessageAsync(recepient, img);
                return true;
            }
            catch (Exception) { return false; }
            finally
            {
                ChatMessage msg = new ChatMessage { Author = UserName, Picture = pic, Time = DateTime.Now, IsOriginNative = true };
                SelectedParticipant.Chatter.Add(msg);
            }           
        }

        private bool CanSendImageMessage()
        {
            return (IsConnected && _selectedParticipant != null && _selectedParticipant.IsLoggedIn);
        }
        #endregion

        #region Select Profile Picture Command
        private ICommand _selectProfilePicCommand;
        public ICommand SelectProfilePicCommand
        {
            get
            {
                return _selectProfilePicCommand ?? (_selectProfilePicCommand =
                    new RelayCommand((o) => SelectProfilePic()));
            }
        }

        private void SelectProfilePic()
        {
            var pic = dialogService.OpenFile("Select image file", "Images (*.jpg;*.png)|*.jpg;*.png");
            if (!string.IsNullOrEmpty(pic))
            {
                var img = Image.FromFile(pic);
                if (img.Width > MAX_IMAGE_WIDTH || img.Height > MAX_IMAGE_HEIGHT)
                {
                    dialogService.ShowNotification($"Image size should be {MAX_IMAGE_WIDTH} x {MAX_IMAGE_HEIGHT} or less.");
                    return;
                }
                ProfilePic = pic;
            }
        }
        #endregion

        #region Open Image Command
        private ICommand _openImageCommand;
        public ICommand OpenImageCommand
        {
            get
            {
                return _openImageCommand ?? (_openImageCommand =
                    new RelayCommand<ChatMessage>((m) => OpenImage(m)));
            }
        }

        private void OpenImage(ChatMessage msg)
        {
            var img = msg.Picture;
            if (string.IsNullOrEmpty(img) || !File.Exists(img)) return;
            Process.Start(img);
        }
        #endregion

        #region Event Handlers
        private void NewTextMessage(string name, string msg, MessageType mt)
        {
            if (mt == MessageType.Unicast)
            {
                ChatMessage cm = new ChatMessage { Author = name, Message = msg, Time = DateTime.Now };
                var sender = _participants.Where((u) => string.Equals(u.Name, name)).FirstOrDefault();
                ctxTaskFactory.StartNew(() => sender.Chatter.Add(cm)).Wait();

                if (!(SelectedParticipant != null && sender.Name.Equals(SelectedParticipant.Name)))
                {
                    ctxTaskFactory.StartNew(() => sender.HasSentNewMessage = true).Wait();
                }
            }
        }

        private void NewImageMessage(string name, byte[] pic, MessageType mt)
        {
            if (mt == MessageType.Unicast)
            {
                var imgsDirectory = Path.Combine(Environment.CurrentDirectory, "Image Messages");
                if (!Directory.Exists(imgsDirectory)) Directory.CreateDirectory(imgsDirectory);

                var imgsCount = Directory.EnumerateFiles(imgsDirectory).Count() + 1;
                var imgPath = Path.Combine(imgsDirectory, $"IMG_{imgsCount}.jpg");

                ImageConverter converter = new ImageConverter();
                using (Image img = (Image)converter.ConvertFrom(pic))
                {
                    img.Save(imgPath);
                }

                ChatMessage cm = new ChatMessage { Author = name, Picture = imgPath, Time = DateTime.Now };
                var sender = _participants.Where(u => string.Equals(u.Name, name)).FirstOrDefault();
                ctxTaskFactory.StartNew(() => sender.Chatter.Add(cm)).Wait();

                if (!(SelectedParticipant != null && sender.Name.Equals(SelectedParticipant.Name)))
                {
                    ctxTaskFactory.StartNew(() => sender.HasSentNewMessage = true).Wait();
                }
            }
        }

        private void ParticipantLogin(User u)
        {
            var ptp = Participants.FirstOrDefault(p => string.Equals(p.Name, u.Name));
            if (_isLoggedIn && ptp == null)
            {
                ctxTaskFactory.StartNew(() => Participants.Add(new Participant
                {
                    Name = u.Name,
                    Photo = u.Photo
                })).Wait();
            }
        }

        private void ParticipantDisconnection(string name)
        {
            var person = Participants.Where((p) => string.Equals(p.Name, name)).FirstOrDefault();
            if (person != null) person.IsLoggedIn = false;
        }

        private void ParticipantReconnection(string name)
        {
            var person = Participants.Where((p) => string.Equals(p.Name, name)).FirstOrDefault();
            if (person != null) person.IsLoggedIn = true;
        }

        private void Reconnecting()
        {
            IsConnected = false;
            IsLoggedIn = false;
        }

        private async void Reconnected()
        {
            var pic = Avatar();
            if (!string.IsNullOrEmpty(_userName)) await chatService.LoginAsync(_userName, pic);
            IsConnected = true;
            IsLoggedIn = true;
        }

        private async void Disconnected()
        {
            var connectionTask = chatService.ConnectAsync();
            await connectionTask.ContinueWith(t => {
                if (!t.IsFaulted)
                {
                    IsConnected = true;
                    chatService.LoginAsync(_userName, Avatar()).Wait();
                    IsLoggedIn = true;
                }
            });
        }

        private void ParticipantTyping(string name)
        {
            var person = Participants.Where((p) => string.Equals(p.Name, name)).FirstOrDefault();
            if (person != null && !person.IsTyping)
            {
                person.IsTyping = true;
                Observable.Timer(TimeSpan.FromMilliseconds(1500)).Subscribe(t => person.IsTyping = false);
            }
        }
        #endregion

        private byte[] Avatar()
        {
            byte[] pic = null;
            if (!string.IsNullOrEmpty(_profilePic)) pic = File.ReadAllBytes(_profilePic);
            return pic;
        }

        public MainWindowViewModel(IChatService chatSvc, IDialogService diagSvc)
        {
            dialogService = diagSvc;
            chatService = chatSvc;

            chatSvc.NewTextMessage += NewTextMessage;
            chatSvc.NewImageMessage += NewImageMessage;
            chatSvc.ParticipantLoggedIn += ParticipantLogin;
            chatSvc.ParticipantLoggedOut += ParticipantDisconnection;
            chatSvc.ParticipantDisconnected += ParticipantDisconnection;
            chatSvc.ParticipantReconnected += ParticipantReconnection;
            chatSvc.ParticipantTyping += ParticipantTyping;
            chatSvc.ConnectionReconnecting += Reconnecting;
            chatSvc.ConnectionReconnected += Reconnected;
            chatSvc.ConnectionClosed += Disconnected;

            ctxTaskFactory = new TaskFactory(TaskScheduler.FromCurrentSynchronizationContext());
        }

    }
}